lib/api_watch_diff.ml

```diff
@@ -1,12 +1,74 @@
+open Types
+
+type 'a change_type = Added | Removed | Modified of 'a
+
+type diff =
+  | Value of string * (value_description * value_description) change_type
+  | Any
+
+type kind = Val
+type name_kind = string * kind
+
+module FieldMap = Map.Make (struct
+  type t = name_kind
+
+  let compare = Stdlib.compare
+end)
+
+let rec extract_values tbl = function
+  | [] -> tbl
+  | Sig_value (id, val_des, _) :: rem ->
+      extract_values (FieldMap.add (Ident.name id, Val) val_des tbl) rem
+  | _ :: rem -> extract_values tbl rem
+
+let compare_values ~reference ~current =
+  let env = Env.empty in
+  let typing_env = Env.add_signature reference (Env.in_signature true env) in
+  let typing_env = Env.add_signature current typing_env in
+  let ref_values = extract_values FieldMap.empty reference in
+  let curr_values = extract_values FieldMap.empty current in
+  let diffs = ref [] in
+  FieldMap.iter
+    (fun name curr_vd ->
+      match name with
+      | val_name, _ ->
+          if FieldMap.mem name ref_values then
+            let ref_vd = FieldMap.find name ref_values in
+
+            let val_coercion () =
+              Includecore.value_descriptions ~loc:ref_vd.val_loc typing_env
+                val_name curr_vd ref_vd
+            in
+            match val_coercion () with
```

Comment:
I should add the whole logic which detects changes (adding, removing and modifying) of a value to `diff_logic` right? 
Also, if I move the whole logic there, I don't think a `unit option` would work, as we are using the `diff list` in our expect tests to print the changes to values.

Or are you suggesting me to move just the logic for detecting modifications there and then returning a bool for that?

Reply:
I was suggesting to extract a function that compares the two matching value descriptions. The detection of addition and removal has to be done at a higher level, when comparing the entire set of values.

E.g. extract something like:
```ocaml
let diff_value ~typing_env vd_ref vd_cur =
  let val_coercion () =
    Includecore.value_descriptions ~loc:ref_vd.val_loc typing_env
      val_name curr_vd ref_vd
  in
  match val_coercion () with
  | Tcoerce_none -> None
  | _ -> Some ()
  | exception Includecore.Dont_match _ -> Some ()
```

In the future, in the diff case, we'll return a concrete diff type instead of `()`.